File: python/leetcode017.py
map = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl', '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}

class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        l = len(digits)
        if l == 0:
            return []
        if l == 1:
            return [x for x in map[digits[0]]]
        else:
            res = []
            s = map[digits[0]]
            for x in s:
                temp = self.letterCombinations(digits[1:])
                for t in temp:
                    res.append(x+t)
            return res


if __name__ == "__main__":
    s = Solution()
    print(s.letterCombinations('23'))

File: leetcodec++/leetcode017.cpp
//
// Created by 黄敏辉 on 2019-06-27.
//

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.size() == 0)
            return res;
        static const vector<string> keyboard{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        res.push_back("");
        for (int i = 0; i < digits.size(); i++){
            const string& candidate = keyboard[digits[i] - '0'];
            if(candidate.empty())
                continue;
            vector<string> tmp;
            for (int j = 0; j < candidate.size(); j++)
                for (int k = 0; k < res.size(); k++)
                    tmp.push_back(res[k] + candidate[j]);
            res.swap(tmp);
        }
        return res;
    }
};